test/src/View/GridTest.cpp

```diff
@@ -147,5 +155,75 @@ namespace TrenchBroom {
         }
+        
+        TEST(GridTest, moveDeltaForPoint) {
+            const auto grid16 = Grid(4);
+            
+            const auto pointOffGrid = Vec3d(17, 17, 17);
+            const auto inputDelta = Vec3d(1, 1, 7); // moves point to (18, 18, 24)
+            const auto pointOnGrid = Vec3d(17, 17, 32);
+
+            ASSERT_EQ(pointOnGrid, pointOffGrid + grid16.moveDeltaForPoint(pointOffGrid, worldBounds, inputDelta));
+        }
+        
+        TEST(GridTest, moveDeltaForPoint_SubInteger) {
+            const auto grid05 = Grid(-1);
+            
+            const auto pointOffGrid = Vec3d(0.51, 0.51, 0.51);
+            const auto inputDelta = Vec3d(0.01, 0.01, 0.30); // moves point to (0.52, 0.52, 0.81)
+            const auto pointOnGrid = Vec3d(0.51, 0.51, 1.0);
+            
+            ASSERT_EQ(pointOnGrid, pointOffGrid + grid05.moveDeltaForPoint(pointOffGrid, worldBounds, inputDelta));
+        }
+        
+        TEST(GridTest, moveDeltaForPoint_SubInteger2) {
+            const auto grid05 = Grid(-1);
+            
+            const auto pointOffGrid = Vec3d(0.51, 0.51, 0.51);
+            const auto inputDelta = Vec3d(0.01, 0.01, 1.30); // moves point to (0.52, 0.52, 1.81)
+            const auto pointOnGrid = Vec3d(0.51, 0.51, 2.0);
+            
+            ASSERT_EQ(pointOnGrid, pointOffGrid + grid05.moveDeltaForPoint(pointOffGrid, worldBounds, inputDelta));
+        }
+        
+        static Model::Brush* makeCube128() {
+            Assets::Texture texture("testTexture", 64, 64);
+            Model::World world(Model::MapFormat::Standard, nullptr, worldBounds);
+            Model::BrushBuilder builder(&world, worldBounds);
+            Model::Brush* cube = builder.createCube(128.0, "");
+            return cube;
+        }
+        
+        TEST(GridTest, moveDeltaForFace) {
+            const auto grid16 = Grid(4);
+            
+            Model::Brush* cube = makeCube128();
+            Model::BrushFace* topFace = cube->findFace(Vec3::PosZ);
+            
+            ASSERT_FLOAT_EQ(64.0, topFace->boundsCenter().z());
+    
+            // try to move almost 4 grid increments up -> snaps to 3
+            ASSERT_EQ(Vec3(0,0,48), grid16.moveDelta(topFace, Vec3(0, 0, 63)));
+            ASSERT_EQ(Vec3(0,0,64), grid16.moveDelta(topFace, Vec3(0, 0, 64)));
+            ASSERT_EQ(Vec3(0,0,64), grid16.moveDelta(topFace, Vec3(0, 0, 65)));
+            
+            delete cube;
+        }
+        
+        TEST(GridTest, moveDeltaForFace_SubInteger) {
+            const auto grid05 = Grid(-1);
+            
+            Model::Brush* cube = makeCube128();
+            Model::BrushFace* topFace = cube->findFace(Vec3::PosZ);
+            
+            ASSERT_FLOAT_EQ(64.0, topFace->boundsCenter().z());
+            
+            // try to move almost 4 grid increments up -> snaps to 3
+            ASSERT_EQ(Vec3(0,0,1.5), grid05.moveDelta(topFace, Vec3(0, 0, 1.9)));
```

Comment:
@kduske mind checking if this test expectation is correct?
This line fails - it only moves 2 grid increments up (1.0 units in +Z) whereas I was expecting 3 grid increments, based on the behaviour of the previous test case for grid 64.

Reply:
Let's look at the computation:

`normDelta.dot(normDelta.firstAxis())` yields `1.9`, this is rounded down to `1` by the inner static cast. Dividing this by the grid size `0.5` yields 2, which is again static_casted without any effect. So by the computation, this is correct. This is then decremented by 1, so it will skip one grid plane and snap to the next. This explains the result you're seeing.

I think for the sub integer grid, we may have to actually round the value as opposed to always rounding it towards 0. But I don't know if that will still be correct for integer grids. But then I think it may not have any noticeable effect at all. May be worth a try.

Reply:
> Try to replace the inner static_cast by Math::round and see what happens.

This rounded `1.9` to `2`, then after dividing by `0.5` gives `4` as the value of `gridSkip`.
The overall output of the `grid05.moveDelta(topFace, Vec3(0, 0, 1.9))` call was `(0, 0, 2)`, i.e. 4 grid steps.
Maybe this would be OK, but it's inconsistent with the behaviour of integer grids.

Instead, I tried deleting the inner static cast, so the gridSkip calculation looks like:
```
            /**
             * Scalar projection of normDelta onto the nearest axial normal vector.
             */
            const FloatType normDeltaScalarProj = normDelta.dot(normDelta.firstAxis());
            
            size_t gridSkip = static_cast<size_t>(normDeltaScalarProj / actualSize());
```
This sets gridSkip to `floor(1.9 / 0.5) = floor(3.8) = 3` which gives the same results as the integer grid case. Seems reasonable,.. does that make sense?

Reply:
Isn‘t that the original formula?

Reply:
No, the original has `static_cast<size_t>` around `normDelta.dot(normDelta.firstAxis())` so that was effectively `floor()`ed before dividing by the grid actual size

Reply:
But I think this will behave very similarly for larger int grid sizes, so it should be okay. I don‘t have a chance to test right now. Does it feel okay when you resize an oddly shaped brush interactively at both large and small grid size?